Added properties and metadata guides (#2468)
deepchecks/deepchecks: docs/source/user-guide/nlp/nlp_metadata.rst

.. _nlp_metadata_guide:

=================
NLP Metadata
=================

Metadata is any additional information about the texts, that can be used to better understand the data and the model's

> behavior. As opposed to the unstructured text, metadata is structured, and is expected to be a pandas DataFrame.

Reviewer: ```suggestion
behavior. In contrast to unstructured text, metadata is structured, and is expected to be a pandas DataFrame.
```

Metadata can be anything - from the source of the data, the timestamp of its creation, to the age of the author, and
any other information that is given about the text.

What Is Metadata Used For?
=============================

Metadata is used by Deepchecks' checks to extract meaningful features from the data, since some computations cannot be

> computed directly on the text (for example, drift). Inspecting the distribution of a metadata column's values (e.g. to

Reviewer: cannot be computed directly on the text - not so clear what is the relation between that fact and the fact we ask for metadata. 
Would perhaps give weak segments as the first example here, as it's clearer and gives a better sense of "why should I care"

Author: yeah, it remained from properties accidentally. I changed the sentence a bit, tell me WDYT

different metadata columns. The number of rows in the metadata DataFrame must be equal to the number of samples in the
dataset, and the order of the rows must be the same as the order of the samples in the dataset.

How To Pass Metadata To Deepchecks
-----------------------------------
In order to pass metadata to Deepchecks, you must pass it to the :class:`TextData <deepchecks.nlp.TextData>` object.
The metadata can be passed to the :class:`TextData <deepchecks.nlp.TextData>` object in two ways:

#. When initializing the :class:`TextData <deepchecks.nlp.TextData>` object, by passing the metadata to the
   ``metadata`` parameter.
#. After initializing the :class:`TextData <deepchecks.nlp.TextData>` object, by using the ``TextData.set_metadata``
   function.

In both options, it's advised to also use the ``categorical_metadata`` parameter to specify which metadata columns are
categorical. The parameter should be a list of the names of the categorical columns.

In the following example, we will pass the metadata to the ``TextData`` object in order to use the
``MetadataSegmentsPerformance`` check:

> .. code-block:: python

Reviewer: Consider showing a printout of a metadata dataframe

Author: This is an rst file.

Reviewer: Can add a table to rst

Author: Seems weird and unnecessary IMO... Is it important to you?

Reviewer: don't invest in it more than 5 minutes, but I bet you gpt4 can do that in 1.

Author: My problem is not that it would take time, it's that I don't think it would look good.

> .. _nlp_metadata_guide:

Reviewer 2: ```suggestion
.. _nlp__metadata_guide:
```

=================
NLP Metadata
=================

Metadata is any additional information about the texts, that can be used to better understand the data and the model's
behavior. In contrast to unstructured text, metadata is structured, and is expected to be a pandas DataFrame.
Metadata can be anything - from the source of the data, the timestamp of its creation, to the age of the author, and
any other information that is given about the text.

Not to be confused with :ref:`text properties <nlp_properties_guide>`, which are features extracted from the text
itself, such as text length or sentiment.

What Is Metadata Used For?
=============================

Metadata is used by Deepchecks' checks to examine the dataset, without directly using the text, as some tests are
difficult to perform on unstructured data.

> Inspecting the distribution of a metadata column's values (e.g. to notice that the source of the texts is different

Reviewer: Still think the weak segment performance example is better as a first sentence because we don't do metadata drift, but can let this go.

Author: I think it's more complicated to explain in a short parenthesis, that's why I kept this example.
